### DISCUSSION TITLE
trojan forward or v2ray dokodemo mode support?!

### DISCUSSION THREAD

**Discussion by mokhtarabadi** [eycorsican/leaf#46] [Q&A]
I want to forward the local port 53 to 1.1.1.1:53 by using trojan or vmess/vless, but I don't know how to that?
like forward mode in trojan-gfw or dokodemo in v2ray

**Comment by mokhtarabadi** [comment]
i see the dns client code, how that works?
i need to add a local dns server for android tun2socks which resolved  hostnames by sending queries to the server through outbounds

**Comment by eycorsican** [comment]
leaf doesn't have a dns server feature.

**Reply by mokhtarabadi** [reply]
I ported it on android, everything is ready the problem only is WebSocket, when I want to use WebSocket I need to set the IP of the server instead of the server domain name (and because the Nginx didn't work because of that need Host header), and leaf hangs because can't resolve the WebSocket host (due to android VPN service is on, loop!)

and or the second scenario is, can I call protect the connection fd in dns_client (which I think set in dns_Servers in the JSON config), if I do it, the resolving of the WebSocket will be work?

**Reply by eycorsican** [reply]
there's also a global function for creating UDP sockets, where you could do the `protect` step: https://github.com/eycorsican/leaf/blob/00ce9b83a418ea675fa5d6e7f4f89bd6ade40f9c/leaf/src/proxy/mod.rs#L102

**Reply by mokhtarabadi** [reply]
So, the only thing I need to protect global TCP, UDP, it's true?

I'm just waiting for websocket custom headers implementation

**Reply by eycorsican** [reply]
the dns client uses the same function for creating sockets when sending dns queries: https://github.com/eycorsican/leaf/blob/00ce9b83a418ea675fa5d6e7f4f89bd6ade40f9c/leaf/src/app/dns_client.rs#L92

so i think yes, you just need to  protect the socket in this function.

well, i will have a try later to support the custom headers.
